masonry/src/passes/mutate.rs

```diff
@@ -0,0 +1,56 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use tracing::info_span;
+
+use crate::passes::merge_state_up;
+use crate::render_root::RenderRoot;
+use crate::widget::WidgetMut;
+use crate::{MutateCtx, Widget, WidgetId, WidgetState};
+
+pub(crate) fn mutate_widget(
+    root: &mut RenderRoot,
+    id: WidgetId,
+    mutate_fn: impl FnOnce(WidgetMut<'_, Box<dyn Widget>>),
+) {
+    let mut dummy_state = WidgetState::root(root.root.id(), root.get_kurbo_size());
+    let (widget_mut, state_mut) = root.widget_arena.get_pair_mut(id);
+
+    let _span = info_span!("mutate_widget").entered();
+    let root_widget = WidgetMut {
+        ctx: MutateCtx {
+            global_state: &mut root.state,
+            // FIXME - This works in practice, because the loop below will merge the
+            // state up to the root. But it's semantically awkward.
+            // parent_widget_state should probably be an Option.
+            parent_widget_state: &mut dummy_state,
```

Comment:
I think the move to `Option` should happen in this PR. AIUI, the only thing this impacts is the `Drop` implementation, right? So it should be a relatively undisruptive change.